src/indexing.jl

```diff
@@ -0,0 +1,399 @@
+  @_propagate_inbounds_meta
+  style = IndexStyle(A)
+  # canonical linear indexing
+  if style == IndexLinear() && length(I) == 1
+    @boundscheck checkbounds(A, I...)
+    return only(I) in eachstoredindex(style, A)
+  end
+
+  # canonical cartesian indexing
+  if style == IndexCartesian() && length(I) == ndims(A)
+    @boundscheck checkbounds(A, I...)
+    return CartesianIndex(I...) in eachstoredindex(style, A)
+  end
+
+  # non-canonical indexing
+  return _isstored(style, A, Base.to_indices(A, I)...)
+end
+
+@interface ::AbstractSparseArrayInterface function getunstoredindex(
+  A::AbstractArray, I::Int...
+)
+  @_propagate_inbounds_meta
+  style = IndexStyle(A)
+
+  # canonical linear indexing
+  if style == IndexLinear() && length(I) == 1
+    @boundscheck checkbounds(A, I...)
+    return zero(eltype(A))
+  end
+
+  # canonical cartesian indexing
+  if style == IndexCartesian() && length(I) == ndims(A)
+    @boundscheck checkbounds(A, I...)
+    return zero(eltype(A))
+  end
+
+  # non-canonical indexing
+  return _getunstoredindex(style, A, Base.to_indices(A, I)...)
+end
+
+# make sure we don't call AbstractArrayInterface defaults
+@interface ::AbstractSparseArrayInterface function getstoredindex(
+  A::AbstractArray, I::Int...
+)
+  @_propagate_inbounds_meta
+  style = IndexStyle(A)
+  error_if_canonical_getstoredindex(style, A, I...)
+  return _getstoredindex(style, A, Base.to_indices(A, I)...)
+end
+
+for f! in (:setstoredindex!, :setunstoredindex!)
+  _f! = Symbol(:_, f!)
+  error_if_canonical_setstoredindex = Symbol(:error_if_canonical_, f!)
+  @eval begin
+    @interface ::AbstractSparseArrayInterface function $f!(A::AbstractArray, v, I::Int...)
+      @_propagate_inbounds_meta
+      style = IndexStyle(A)
+      $error_if_canonical_setstoredindex(style, A, I...)
+      return $_f!(style, A, v, Base.to_indices(A, I)...)
+    end
+  end
+end
+
+@interface ::AbstractSparseArrayInterface storedlength(A::AbstractArray) =
+  length(storedvalues(A))
+@interface ::AbstractSparseArrayInterface storedpairs(A::AbstractArray) =
+  Iterators.map(=>, eachstoredindex(A), storedvalues(A))
+
+#=
+All sparse array interfaces are mapped through layout_getindex. (is this too opinionated?)
+
+using ArrayLayouts getindex: this is a bit cumbersome because there already is a way to make
+that work focused on types but here we want to focus on interfaces.
+eg: ArrayLayouts.@layoutgetindex ArrayType
+TODO: decide if we need the interface approach at all here
+=#
+for (Tr, Tc) in Iterators.product(
+  Iterators.repeated((:Colon, :AbstractUnitRange, :AbstractVector, :Integer), 2)...
+)
+  Tr === Tc === :Integer && continue
+  @eval begin
+    @interface ::AbstractSparseArrayInterface function Base.getindex(
+      A::AbstractMatrix, kr::$Tr, jr::$Tc
+    )
+      Base.@inline # needed to make boundschecks work
+      return ArrayLayouts.layout_getindex(A, kr, jr)
+    end
+  end
+end
```

Comment:
Maybe it is better to avoid calling out to `layout_getindex`, what do you think?